nni/algorithms/hpo/gridsearch_tuner.py

```diff
@@ -2,209 +2,241 @@
+        if params is None:
+            raise nni.NoMoreTrialError('Search space fully explored')
+        return deformat_parameters(params, self.space)
+
+    def receive_trial_result(self, *args, **kwargs):
+        pass
 
-    Type ``choice`` will select one of the options. Note that it can also be nested.
+    def import_data(self, data):  # for resuming
+        # this is not accurate for concurrency > 1
+        # but I believe it's not worthy to introduce complex logic to fix this
+        # instead we should support checkpoint (TODO)
+        for _ in data:
+            self._suggest()
 
-    Type ``quniform`` will receive three values [``low``, ``high``, ``q``],
-    where [``low``, ``high``] specifies a range and ``q`` specifies the interval.
-    It will be sampled in a way that the first sampled value is ``low``,
-    and each of the following values is 'interval' larger than the value in front of it.
+    def _suggest(self):
+        # returns next parameter set, or None if the space is already fully explored
+        while True:
+            if self.grid is None:  # search space fully explored
+                return None
 
-    Type ``randint`` gives all possible intergers in range[``low``, ``high``). Note that ``high`` is not included.
-    """
+            self._next_vector()
 
-    def __init__(self):
-        self.count = -1
-        self.expanded_search_space = []
-        self.supplement_data = dict()
-
-    def _json2parameter(self, ss_spec):
-        """
-        Generate all possible configs for hyperparameters from hyperparameter space.
-
-        Parameters
-        ----------
-        ss_spec : dict or list
-            Hyperparameter space or the ``_value`` of a hyperparameter
-
-        Returns
-        -------
-        list or dict
-            All the candidate choices of hyperparameters. for a hyperparameter, chosen_params
-            is a list. for multiple hyperparameters (e.g., search space), chosen_params is a dict.
-        """
-        if isinstance(ss_spec, dict):
-            if '_type' in ss_spec.keys():
-                _type = ss_spec['_type']
-                _value = ss_spec['_value']
-                chosen_params = list()
-                if _type == 'choice':
-                    for value in _value:
-                        choice = self._json2parameter(value)
-                        if isinstance(choice, list):
-                            chosen_params.extend(choice)
-                        else:
-                            chosen_params.append(choice)
-                elif _type == 'quniform':
-                    chosen_params = self._parse_quniform(_value)
-                elif _type == 'randint':
-                    chosen_params = self._parse_randint(_value)
-                else:
-                    raise RuntimeError("Not supported type: %s" % _type)
+            if self.vector is None:  # epoch end, update grid and retry
+                self._next_grid()
```

Comment:
if there is a uniform, this grid search will never end, right?

Reply:
Yes.